soweego/commons/url_utils.py

```diff
@@ -209,57 +207,94 @@ def tokenize(url, domain_only=False) -> set:
 
 
 def get_external_id_from_url(url, ext_id_pids_to_urls):
-    LOGGER.debug('Trying to extract an identifier from URL <%s>', url)
-    url = url.rstrip('/')
-    # Always use HTTPS
-    if not url.startswith('https'):
-        url = url.replace('http', 'https')
+    LOGGER.debug('Trying to extract an identifier from <%s>', url)
+
+    # Tidy up: remove trailing slash & use HTTPS
+    tidy = url.rstrip('/')
+    if not tidy.startswith('https'):
+        tidy = tidy.replace('http', 'https')
+
+    # Start extraction
     for pid, formatters in ext_id_pids_to_urls.items():
-        for formatter_url, formatter_regex in formatters.items():
+        for formatter_url, (id_regex, url_regex,) in formatters.items():
+
+            # Optimal case: match the original input URL against a full URL regex
+            if url_regex is not None:
+# TODO refactor this block, duplicate with the next
+                match = (
+                    re.match(url_regex, url)
+                    if isinstance(url_regex, re.Pattern)
+                    else regex.match(url_regex, url)
+                )
+                if match is not None:
+                    groups = match.groups()
+                    # This shouldn't happend, but who knows?
+                    if len(groups) > 1:
+                        LOGGER.debug(
```

Comment:
better to have it as a warning since it "shouldn't happen" - check logs later

Reply:
```suggestion
                        LOGGER.warning(
```